### PR TITLE
SNOW-3022768: ODBC API tests for connection termination functions

### PR DESCRIPTION
**Pull request by sfc-gh-ddas**
Added granular ODBC unit and integration tests to verify compliance of ODBC wrapper with the ODBC 3.x and 2.x specs and consistent behaviour with existing ODBC driver.

This PR only contains ["Terminating a connection" category functions](https://docs.google.com/spreadsheets/d/1BjOHFnLQ2AooXGqWncsaqCLIw8z9XD0v3fwwF-S-E3g/edit?gid=999869428#gid=999869428).

### CODE REVIEW

File: odbc_tests/tests/odbc-api/terminating_connection/free_handle_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+                             SQL_NTS, nullptr, 0, nullptr, 0);
+  REQUIRE(ret == SQL_SUCCESS);
+
+  // Allocate multiple statements
+  SQLHSTMT stmt1 = SQL_NULL_HSTMT, stmt2 = SQL_NULL_HSTMT, stmt3 = SQL_NULL_HSTMT;
+  ret = SQLAllocHandle(SQL_HANDLE_STMT, dbc, &stmt1);
+  REQUIRE(ret == SQL_SUCCESS);
+  ret = SQLAllocHandle(SQL_HANDLE_STMT, dbc, &stmt2);
+  REQUIRE(ret == SQL_SUCCESS);
+  ret = SQLAllocHandle(SQL_HANDLE_STMT, dbc, &stmt3);
```
Comment (sfc-gh-makowalski) [inline odbc_tests/tests/odbc-api/terminating_connection/free_handle_tests.cpp]: why do we use `dbc` here instead of `dbc_handle`?

File: odbc_tests/tests/odbc-api/terminating_connection/disconnect_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+// ============================================================================
+// SQLDisconnect - Edge Cases
+// ============================================================================
+
+TEST_CASE_METHOD(DbcDefaultDSNFixture, "SQLDisconnect: After failed connection attempt",
+                 "[odbc-api][disconnect][terminating_connection]") {
+  SKIP_NEW_DRIVER_NOT_IMPLEMENTED();
+
+  // Attempt to connect with invalid credentials
+  SQLRETURN   ret = SQLConnect(dbc_handle(), reinterpret_cast<SQLCHAR*>(const_cast<char*>("InvalidDSN")), SQL_NTS, nullptr, 0, nullptr, 0);
```
Comment (sfc-gh-makowalski) [inline odbc_tests/tests/odbc-api/terminating_connection/disconnect_tests.cpp]: too many spaces

File: odbc_tests/tests/odbc-api/terminating_connection/disconnect_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,315 @@
+#include <sql.h>
+#include <sqlext.h>
+#include <sqltypes.h>
+
+#include <catch2/catch_test_macros.hpp>
+#include <catch2/matchers/catch_matchers_all.hpp>
```
Comment (sfc-gh-makowalski) [inline odbc_tests/tests/odbc-api/terminating_connection/disconnect_tests.cpp]: dead import?

File: odbc_tests/tests/odbc-api/terminating_connection/disconnect_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+#include "compatibility.hpp"
+#include "Connection.hpp"
+#include "ODBCConfig.hpp"
+#include "ODBCFixtures.hpp"
+#include "get_diag_rec.hpp"
+#include "test_macros.hpp"
+#include "test_setup.hpp"
+
+using namespace Catch::Matchers;
```
Comment (sfc-gh-makowalski) [inline odbc_tests/tests/odbc-api/terminating_connection/disconnect_tests.cpp]: dead import?

File: odbc_tests/tests/odbc-api/terminating_connection/free_handle_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+  REQUIRE(ret == SQL_SUCCESS);
+
+  SQLHDBC dbc2 = SQL_NULL_HDBC;
+  ret = SQLAllocHandle(SQL_HANDLE_DBC, env_handle(), &dbc2);
+  REQUIRE(ret == SQL_SUCCESS);
+
+  // Note: Reference driver returns SQL_ERROR when getting attribute from unconnected handle
+  SQLUINTEGER timeout = 999;
+  ret = SQLGetConnectAttr(dbc2, SQL_ATTR_CONNECTION_TIMEOUT, &timeout, 0, nullptr);
+  REQUIRE(ret == SQL_ERROR);
```
Comment (sfc-gh-makowalski) [inline odbc_tests/tests/odbc-api/terminating_connection/free_handle_tests.cpp:540]: I think we should check if its the default rather than checking for error.

File: odbc_tests/tests/odbc-api/terminating_connection/free_handle_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,546 @@
+#include <sql.h>
+#include <sqlext.h>
+#include <sqltypes.h>
+
+#include <catch2/catch_test_macros.hpp>
+#include <catch2/matchers/catch_matchers_all.hpp>
```
Comment (sfc-gh-makowalski) [inline odbc_tests/tests/odbc-api/terminating_connection/free_handle_tests.cpp]: is this used?

File: odbc_tests/tests/odbc-api/terminating_connection/free_handle_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+
+#include "compatibility.hpp"
+#include "Connection.hpp"
+#include "ODBCConfig.hpp"
+#include "ODBCFixtures.hpp"
+#include "get_diag_rec.hpp"
+#include "test_macros.hpp"
+#include "test_setup.hpp"
+
+using namespace Catch::Matchers;
```
Comment (sfc-gh-makowalski) [inline odbc_tests/tests/odbc-api/terminating_connection/free_handle_tests.cpp]: is this used?

**Review by sfc-gh-makowalski — CHANGES_REQUESTED** [review]

File: odbc_tests/tests/odbc-api/terminating_connection/disconnect_tests.cpp (hunk 3 above)
Comment (sfc-gh-ddas) [reply odbc_tests/tests/odbc-api/terminating_connection/disconnect_tests.cpp]: You're right, the matchers are not used.

**Review by sfc-gh-jszczerbinski — APPROVED** [review]

File: odbc_tests/tests/odbc-api/terminating_connection/free_handle_tests.cpp (hunk 5 above)
Comment (sfc-gh-ddas) [reply odbc_tests/tests/odbc-api/terminating_connection/free_handle_tests.cpp:540]: Your right, added a second connection to check.

**Review by sfc-gh-makowalski — APPROVED** [review]
